### PR TITLE
Enabling conditional query

### PR DESCRIPTION
**Pull request by TsuyoshiUshio**
Hi @cgillum , 

I develop the conditional query feature for the durable functions. 
Could you review this, please?

# Related issue and PR

This PR is related 

https://github.com/Azure/durabletask/pull/200
https://github.com/Azure/azure-functions-durable-extension/issues/358

# TODO 

For the latest version of the Durable Extension uses DurableTask.AzureStorage.Private v1.2.4 However, current DurableTask.AzureStorage is 1.2.3
Maybe you might implement some improvement. So I don't include csproj file.  Before starting CI, we might need to publish DurableTask.AzureStorage new version for CI testing. 

I don't implement pager feature. (It is used inside of the DurableTask, however, not for OrchestrationClient and HttpApi now. It will be the separate PR.

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs [hunk 1]
```diff
@@ -210,5 +210,16 @@ public virtual Task<DurableOrchestrationStatus> GetStatusAsync(string instanceId
...
+
+        /// <summary>
+        /// Gets all the status which match the conditions.
+        /// </summary>
+        /// <param name="createdTimeFrom">fetch status greater than the createdTime.</param>
+        /// <param name="createdTimeTo">fetch status less than the createdTime.</param>
+        /// <param name="runtimeStatus">fetch status match the runtimeStatus</param>
+        /// <param name="cancellationToken">Cancellation token that can be used to cancel the status query operation.</param>
+        /// <returns>Returns orchestration status for all instances.</returns>
+        public abstract Task<IList<DurableOrchestrationStatus>> GetStatusAsync(DateTime createdTimeFrom, DateTime createdTimeTo, string runtimeStatus, CancellationToken cancellationToken = default(CancellationToken));
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs]: `runtimeStatus` should be a `OrchestrationRuntimeStatus` enumeration, not a string. Also, I think I forgot to ask this in the other PR, but shouldn't this be a list of statuses? For example:

```csharp
public abstract Task<IList<DurableOrchestrationStatus>> GetStatusAsync(
    DateTime createdTimeFrom,
    DateTime? createdTimeTo,
    IEnumerable<OrchestrationRuntimeStatus> runtimeStatus,
    CancellationToken cancellationToken);
```

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs [hunk 2]
```diff
@@ -255,6 +263,14 @@ private StatusResponsePayload ConvertFrom(DurableOrchestrationStatus status)
             };
         }
 
+        private static DateTime GetDateTimeQueryParameterValue(NameValueCollection queryStringNameValueCollection, string queryParameterName)
+        {
+            var value = queryStringNameValueCollection[queryParameterName];
+            DateTime.TryParse(value, out DateTime dateTime);
+            return dateTime;
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs]: It's not clear what the expected `dateTime` value is if `DateTime.TryParse` fails. I suggest you add a `defaultDateTime` parameter to this method and return that if `DateTime.TryParse` returns `false`.

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs [hunk 3]
```diff
@@ -255,6 +263,14 @@ private StatusResponsePayload ConvertFrom(DurableOrchestrationStatus status)
...
         }
 
+        private static DateTime GetDateTimeQueryParameterValue(NameValueCollection queryStringNameValueCollection, string queryParameterName)
+        {
+            var value = queryStringNameValueCollection[queryParameterName];
+            DateTime.TryParse(value, out DateTime dateTime);
+            return dateTime;
+        }
+
+
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs]: nit: 余分な空白を決してください。

File: test/HttpApiHandlerTests.cs [hunk 4]
```diff
@@ -313,6 +314,56 @@ public async Task GetAllStatus_is_Success()
...
+            var actual = JsonConvert.DeserializeObject<IList<StatusResponsePayload>>(await responseMessage.Content.ReadAsStringAsync());
+            clientMock.Verify(x => x.GetStatusAsync(createdTimeFrom, createdTimeTo, runtimeStatus, It.IsAny<CancellationToken>()));
+            Assert.Equal("01", actual[0].InstanceId);
+            Assert.Equal("Running", actual[0].RuntimeStatus);
+            Assert.Equal("02", actual[1].InstanceId);
+            Assert.Equal("Running", actual[1].RuntimeStatus);
+        }
+
+
+
```
Comment (cgillum) [inline test/HttpApiHandlerTests.cs]: nit: 余分な空白を決してください。

File: test/HttpApiHandlerTests.cs [hunk 5]
```diff
@@ -313,6 +314,56 @@ public async Task GetAllStatus_is_Success()
...
+            var runtimeStatus = "Running";
+            var clientMock = new Mock<DurableOrchestrationClientBase>();
+            clientMock
+                .Setup(x => x.GetStatusAsync(createdTimeFrom, createdTimeTo, runtimeStatus, It.IsAny<CancellationToken>()))
+                .Returns(Task.FromResult(list));
+            var httpApiHandler = new ExtendedHttpApiHandler(clientMock.Object);
+
+            var getStatusRequestUriBuilder = new UriBuilder(TestConstants.NotificationUrl);
+            getStatusRequestUriBuilder.Path += $"/Instances/";
+            getStatusRequestUriBuilder.Query = $"createdTimeFrom={System.Web.HttpUtility.UrlEncode(createdTimeFrom.ToString())}&createdTimeTo={System.Web.HttpUtility.UrlEncode(createdTimeTo.ToString())}&runtimeStatus={runtimeStatus}";
```
Comment (cgillum) [inline test/HttpApiHandlerTests.cs]: `System.Web.HttpUtility.UrlEncode`を使わないでください。`WebUtility.UrlEncode`を使ったほうがいいです。

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs [hunk 6]
```diff
@@ -173,6 +173,21 @@ public override async Task<IList<DurableOrchestrationStatus>> GetStatusAsync(Can
             return results;
         }
 
+        /// <inheritdoc />
+        public override async Task<IList<DurableOrchestrationStatus>> GetStatusAsync(DateTime createdTimeFrom, DateTime createdTimeTo, string runtimeStatus, CancellationToken cancellationToken = default(CancellationToken))
```
[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs]: Done

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs [hunk 7]
```diff
@@ -210,5 +210,16 @@ public virtual Task<DurableOrchestrationStatus> GetStatusAsync(string instanceId
         /// <param name="cancellationToken">Cancellation token that can be used to cancel the status query operation.</param>
         /// <returns>Returns orchestration status for all instances.</returns>
         public abstract Task<IList<DurableOrchestrationStatus>> GetStatusAsync(CancellationToken cancellationToken = default(CancellationToken));
+
+        /// <summary>
+        /// Gets all the status which match the conditions.
```
[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs]: Thanks! Done.

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs [hunk 8]
```diff
@@ -210,5 +210,16 @@ public virtual Task<DurableOrchestrationStatus> GetStatusAsync(string instanceId
...
+        /// </summary>
+        /// <param name="createdTimeFrom">fetch status greater than the createdTime.</param>
```
[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs]: Done

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs [hunk 9]
```diff
@@ -210,5 +210,16 @@ public virtual Task<DurableOrchestrationStatus> GetStatusAsync(string instanceId
...
+        /// <param name="createdTimeTo">fetch status less than the createdTime.</param>
```
[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs]: Done

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs (hunk 1 above)
Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/DurableOrchestrationClientBase.cs]: Done

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs (hunk 2 above)
Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs]: Done

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs (hunk 3 above)
Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs]: Done

Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs]: @cgillum nit: 余分な空白を決してください。 -> 余分な空白を消してください。

File: test/HttpApiHandlerTests.cs (hunk 4 above)
Comment (TsuyoshiUshio) [reply test/HttpApiHandlerTests.cs]: Done

**Comment by TsuyoshiUshio** [comment]
Hi @cgillum, I changed your request and testing now. 
However, when I test HTTP API with createdTimeFrom, it becomes 500.  How can we debug it? I can't see any logs on locally and FunctionApp logs.

**Comment by TsuyoshiUshio** [comment]
Hi @cgillum I finish changing. Could you review this, please?

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs [hunk 10]
```diff
@@ -339,6 +340,19 @@ private async Task<DurableOrchestrationStatus> GetDurableOrchestrationStatusAsyn
...
+        private OrchestrationStatus ConvertFrom(OrchestrationRuntimeStatus runtimeStatus)
+        {
+            OrchestrationStatus status;
+            if (Enum.TryParse<OrchestrationStatus>(runtimeStatus.ToString(), out status))
+            {
+                return status;
+            }
+            else
+            {
+                throw new ArgumentException($"The OrchestrationRuntimeStatus {runtimeStatus} is exptected to be converted into OrchestrationStatus");
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs]: You don't need to convert back and forth from strings. Just do a direct cast like we do in other places:
```csharp
return (OrchestrationStatus)runtimeStatus;
```

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs [hunk 11]
```diff
@@ -26,7 +27,7 @@ internal class HttpApiHandler
         private const string ShowHistoryOutputParameter = "showHistoryOutput";
         private const string CreatedTimeFromParameter = "createdTimeFrom";
         private const string CreatedTimeToParameter = "createdTimeTo";
-        private const string RuntimeStatusParameter = "runtimeStatus";
+        private const string RuntimeStatusParameter = "runtimeStatus[]";
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs]: I don't think we should change this query string parameter name. I have never seen this convention.

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs (hunk 3 above)
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs]: Thanks for the correction. Yes, I meant to say "please delete" and not "please decide" :)

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs [hunk 12]
```diff
@@ -263,13 +265,26 @@ private StatusResponsePayload ConvertFrom(DurableOrchestrationStatus status)
...
         }
 
-        private static DateTime GetDateTimeQueryParameterValue(NameValueCollection queryStringNameValueCollection, string queryParameterName)
+        private static IEnumerable<OrchestrationRuntimeStatus> GetIEnumerableQueryParameterValue(NameValueCollection queryStringNameValueCollection, string queryParameterName)
         {
-            var value = queryStringNameValueCollection[queryParameterName];
-            DateTime.TryParse(value, out DateTime dateTime);
-            return dateTime;
+            var results = new List<OrchestrationRuntimeStatus>();
+            var parameters = queryStringNameValueCollection.GetValues(queryParameterName) ?? new string[] { };
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs]: It looks like you are expecting multiple query string parameters with the same name - e.g. `runtimeStatus[]=foo&runtimeStatus[]=bar`. I do not recommend this design. Instead, I suggest you do it like this `runtimeStatus=foo,bar` - i.e. using a single query string parameter and separating multiple values using a comma. This is more natural, less verbose, and will be better supported across different software stacks. This code will then need to be changed to look for a single parameter value and split it at the comma.

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs (hunk 10 above)
Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs]: Oh. God... I've never known that. Thanks

Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/DurableOrchestrationClient.cs]: Done

File: src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs (hunk 12 above)
Comment (TsuyoshiUshio) [reply src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs]: Totally make sense. I've done it.

**Comment by TsuyoshiUshio** [comment]
If you are OK. I'll start to write document. :)

**Comment by cgillum** [comment]
Yes, that would be great!

**Pull request merged**